File: convert.py
import json

dsl = """
model
  schema 1.1

type organization
  relations
    define member: [user] or owner
    define owner: [user]
    define repo_admin: [user, organization#member]
    define repo_reader: [user, organization#member]
    define repo_writer: [user, organization#member]

type repo
  relations
    define admin: [user, team#member] or repo_admin from owner
    define maintainer: [user, team#member] or admin
    define owner: [organization]
    define reader: [user, team#member] or triager or repo_reader from owner
    define triager: [user] but not repo_reader from owner
    define writer: [user, user:*, team#member] or maintainer or repo_writer from owner
    
type team
  relations
    define member: [user, team#member]

type user
"""


def parse_dsl(dsl):
    lines = dsl.strip().split('\n')
    schema_version = None
    types = []
    current_type = None

    for line in lines:
        line = line.strip()
        if line.startswith('schema'):
            schema_version = line.split()[1]
        elif line.startswith('type'):
            if current_type:
                if not current_type["relations"]:
                    current_type["metadata"] = None
                types.append(current_type)
            current_type = {
                "type": line.split()[1],
                "relations": {},
                "metadata": {
                    "relations": {}
                }
            }
        elif line.startswith('define'):
            parts = line.split(':', 1)  # Split only at the first ':'
            relation_name = parts[0].split()[1]
            relation_def = parts[1].strip()
            current_type["relations"][relation_name] = parse_relation_def(relation_def)
            current_type["metadata"]["relations"][relation_name] = parse_metadata_relation_def(relation_def)
    if current_type:
        if not current_type["relations"]:
            current_type["metadata"] = None
        types.append(current_type)

    return {
        "schema_version": schema_version,
        "type_definitions": types
    }


def parse_relation_def(definition):
    if ' or ' in definition:
        parts = definition.split(' or ')
        return {
            "union": {
                "child": [parse_relation_def(part.strip()) for part in parts]
            }
        }

    if ' and ' in definition:
        parts = definition.split(' and ')
        return {
            "intersection": {
                "child": [parse_relation_def(part.strip()) for part in parts]
            }
        }

    if ' but not ' in definition:
        parts = definition.split(' but not ')
        return {
            "difference": {
                "base": parse_relation_def(parts[0].strip()),
                "subtract": parse_relation_def(parts[1].strip())
            }
        }

    if '[' in definition and ']' in definition:
        return {"this": {}}

    if ' from ' in definition:
        before_from, after_from = definition.split(' from ')
        return {
            "tupleToUserset": {
                "tupleset": {
                    "object": "",
                    "relation": after_from.strip()
                },
                "computedUserset": {
                    "object": "",
                    "relation": before_from.strip()
                }
            }
        }

    return {
        "computedUserset": {
            "object": "",
            "relation": definition
        }
    }


def parse_metadata_relation_def(definition):
    if '[' in definition and ']' in definition:
        start = definition.index("[") + 1
        end = definition.index("]")
        definitions = definition[start:end].split(',')
        directly_related_user_types = []
        for defn in definitions:
            defn = defn.strip()
            if ':' in defn:
                directly_related_user_types.append({"type": defn.split(':')[0], "wildcard": {}})
            elif '#' in defn:
                type_part, relation_part = defn.split('#')
                directly_related_user_types.append({"type": type_part, "relation": relation_part})
            else:
                directly_related_user_types.append({"type": defn})
        return {"directly_related_user_types": directly_related_user_types}
    else:
        return {"directly_related_user_types": []}


dsl_json = parse_dsl(dsl)
print(json.dumps(dsl_json, indent=2))

File: convert.java
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.util.*;

public class convert {

    public static void main(String[] args) {
        String dsl = """
                model
                  schema 1.1

                type organization
                  relations
                    define member: [user] or owner
                    define owner: [user]
                    define repo_admin: [user, organization#member]
                    define repo_reader: [user, organization#member]
                    define repo_writer: [user, organization#member]

                type repo
                  relations
                    define admin: [user, team#member] or repo_admin from owner
                    define maintainer: [user, team#member] or admin
                    define owner: [organization]
                    define reader: [user, team#member] or triager or repo_reader from owner
                    define triager: [user] but not repo_reader from owner
                    define writer: [user, user:*, team#member] or maintainer or repo_writer from owner
                    
                type team
                  relations
                    define member: [user, team#member]

                type user
                """;

        Map<String, Object> dslJson = parseDsl(dsl);
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        String jsonOutput = gson.toJson(dslJson);
        System.out.println(jsonOutput);
    }

    public static Map<String, Object> parseDsl(String dsl) {
        String[] lines = dsl.strip().split("\n");
        String schemaVersion = null;
        List<Map<String, Object>> types = new ArrayList<>();
        Map<String, Object> currentType = null;

        for (String line : lines) {
            line = line.strip();
            if (line.startsWith("schema")) {
                schemaVersion = line.split(" ")[1];
            } else if (line.startsWith("type")) {
                if (currentType != null) {
                    if (((Map<?, ?>) currentType.get("relations")).isEmpty()) {
                        currentType.put("metadata", null);
                    }
                    types.add(currentType);
                }
                currentType = new HashMap<>();
                currentType.put("type", line.split(" ")[1]);
                currentType.put("relations", new HashMap<String, Object>());
                currentType.put("metadata", new HashMap<String, Object>());
                ((Map<String, Object>) currentType.get("metadata")).put("relations", new HashMap<String, Object>());
            } else if (line.startsWith("define")) {
                String[] parts = line.split(":", 2);
                String relationName = parts[0].split(" ")[1];
                String relationDef = parts[1].strip();
                ((Map<String, Object>) currentType.get("relations")).put(relationName, parseRelationDef(relationDef));
                ((Map<String, Object>) ((Map<String, Object>) currentType.get("metadata")).get("relations")).put(relationName, parseMetadataRelationDef(relationDef));
            }
        }

        if (currentType != null) {
            if (((Map<?, ?>) currentType.get("relations")).isEmpty()) {
                currentType.put("metadata", null);
            }
            types.add(currentType);
        }

        Map<String, Object> result = new HashMap<>();
        result.put("schema_version", schemaVersion);
        result.put("type_definitions", types);
        return result;
    }

    public static Map<String, Object> parseRelationDef(String definition) {
        Map<String, Object> result = new HashMap<>();
        if (definition.contains(" or ")) {
            String[] parts = definition.split(" or ");
            List<Map<String, Object>> children = new ArrayList<>();
            for (String part : parts) {
                children.add(parseRelationDef(part.strip()));
            }
            result.put("union", Map.of("child", children));
        } else if (definition.contains(" and ")) {
            String[] parts = definition.split(" and ");
            List<Map<String, Object>> children = new ArrayList<>();
            for (String part : parts) {
                children.add(parseRelationDef(part.strip()));
            }
            result.put("intersection", Map.of("child", children));
        } else if (definition.contains(" but not ")) {
            String[] parts = definition.split(" but not ");
            result.put("difference", Map.of(
                    "base", parseRelationDef(parts[0].strip()),
                    "subtract", parseRelationDef(parts[1].strip())
            ));
        } else if (definition.contains("[") && definition.contains("]")) {
            result.put("this", new HashMap<>());
        } else if (definition.contains(" from ")) {
            String[] parts = definition.split(" from ");
            result.put("tupleToUserset", Map.of(
                    "tupleset", Map.of("object", "", "relation", parts[1].strip()),
                    "computedUserset", Map.of("object", "", "relation", parts[0].strip())
            ));
        } else {
            result.put("computedUserset", Map.of("object", "", "relation", definition));
        }
        return result;
    }

    public static Map<String, Object> parseMetadataRelationDef(String definition) {
        Map<String, Object> result = new HashMap<>();
        List<Map<String, Object>> directlyRelatedUserTypes = new ArrayList<>();
        if (definition.contains("[") && definition.contains("]")) {
            int start = definition.indexOf('[') + 1;
            int end = definition.indexOf(']');
            String[] definitions = definition.substring(start, end).split(",");
            for (String defn : definitions) {
                defn = defn.strip();
                if (defn.contains(":")) {
                    directlyRelatedUserTypes.add(Map.of("type", defn.split(":")[0], "wildcard", new HashMap<>()));
                } else if (defn.contains("#")) {
                    String[] parts = defn.split("#");
                    directlyRelatedUserTypes.add(Map.of("type", parts[0], "relation", parts[1]));
                } else {
                    directlyRelatedUserTypes.add(Map.of("type", defn));
                }
            }
        }
        result.put("directly_related_user_types", directlyRelatedUserTypes);
        return result;
    }
}
